File: sdks/go/webhook.go
package linktor

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"strconv"
	"time"
)

const (
	SignatureHeader = "X-Linktor-Signature"
	TimestampHeader = "X-Linktor-Timestamp"
	DefaultTolerance = 300 // 5 minutes
)

// WebhookEvent represents a webhook event
type WebhookEvent struct {
	ID        string                 `json:"id"`
	Type      string                 `json:"type"`
	Timestamp time.Time              `json:"timestamp"`
	TenantID  string                 `json:"tenantId"`
	Data      map[string]interface{} `json:"data"`
}

// ComputeSignature computes HMAC-SHA256 over the raw payload only. It is the
// low-level primitive; production signatures bind the timestamp too — see
// ComputeTimestampedSignature.
func ComputeSignature(payload []byte, secret string) string {
	h := hmac.New(sha256.New, []byte(secret))
	h.Write(payload)
	return hex.EncodeToString(h.Sum(nil))
}

// ComputeTimestampedSignature computes the HMAC-SHA256 over the Stripe-style
// signing input `timestamp + "." + payload`. This matches what the Linktor
// server places in X-Linktor-Signature, so the unix-seconds X-Linktor-Timestamp
// header is authenticated and cannot be replayed with a fresh timestamp.
func ComputeTimestampedSignature(timestamp string, payload []byte, secret string) string {
	signed := make([]byte, 0, len(timestamp)+1+len(payload))
	signed = append(signed, timestamp...)
	signed = append(signed, '.')
	signed = append(signed, payload...)
	return ComputeSignature(signed, secret)
}

// VerifyWebhookSignature verifies a signature that binds timestamp and payload.
func VerifyWebhookSignature(payload []byte, signature, timestamp, secret string) bool {
	if signature == "" || timestamp == "" || secret == "" {
		return false
	}

	expected := ComputeTimestampedSignature(timestamp, payload, secret)
	return hmac.Equal([]byte(signature), []byte(expected))
}

// VerifyWebhook verifies webhook signature and timestamp freshness. The
// timestamp is mandatory (it is part of the signed material) and stale
// timestamps outside the tolerance window are rejected (anti-replay).
func VerifyWebhook(payload []byte, headers http.Header, secret string, tolerance int) bool {
	if tolerance == 0 {
		tolerance = DefaultTolerance
	}

	signature := headers.Get(SignatureHeader)
	timestampStr := headers.Get(TimestampHeader)
	if signature == "" || timestampStr == "" {
		return false
	}

	ts, err := strconv.ParseInt(timestampStr, 10, 64)
	if err != nil {
		return false
	}
	now := time.Now().Unix()
	if math.Abs(float64(now-ts)) > float64(tolerance) {
		return false
	}

	return VerifyWebhookSignature(payload, signature, timestampStr, secret)
}

// ConstructEvent parses and validates a webhook event
func ConstructEvent(payload []byte, headers http.Header, secret string, tolerance int) (*WebhookEvent, error) {
	if !VerifyWebhook(payload, headers, secret, tolerance) {
		return nil, fmt.Errorf("webhook signature verification failed")
	}

	var event WebhookEvent
	if err := json.Unmarshal(payload, &event); err != nil {
		return nil, fmt.Errorf("invalid JSON payload: %w", err)
	}

	if event.ID == "" || event.Type == "" {
		return nil, fmt.Errorf("invalid webhook event structure")
	}

	return &event, nil
}

// WebhookHandler creates an HTTP handler for webhooks
func WebhookHandler(secret string, handlers map[string]func(*WebhookEvent)) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}

		// Read body
		var payload []byte
		if r.Body != nil {
			var err error
			payload, err = io.ReadAll(r.Body)
			if err != nil {
				http.Error(w, "Failed to read body", http.StatusBadRequest)
				return
			}
		}

		event, err := ConstructEvent(payload, r.Header, secret, 0)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		if handler, ok := handlers[event.Type]; ok {
			handler(event)
		}

		w.WriteHeader(http.StatusOK)
	}
}

File: sdks/rust/src/webhook.rs
use crate::error::{LinktorError, Result};
use crate::types::webhook::{WebhookEvent, SIGNATURE_HEADER, TIMESTAMP_HEADER, DEFAULT_TOLERANCE_SECONDS};
use chrono::Utc;
use hmac::{Hmac, Mac};
use sha2::Sha256;
use std::collections::HashMap;

type HmacSha256 = Hmac<Sha256>;

/// Compute HMAC-SHA256 signature for the given payload
pub fn compute_signature(payload: &[u8], secret: &str) -> String {
    let mut mac = HmacSha256::new_from_slice(secret.as_bytes())
        .expect("HMAC can take key of any size");
    mac.update(payload);
    let result = mac.finalize();
    hex::encode(result.into_bytes())
}

/// Verify webhook signature only (no timestamp validation)
pub fn verify_signature(payload: &[u8], signature: &str, secret: &str) -> bool {
    if signature.is_empty() || secret.is_empty() {
        return false;
    }

    let expected = compute_signature(payload, secret);

    // Constant-time comparison
    if expected.len() != signature.len() {
        return false;
    }

    let mut result = 0u8;
    for (a, b) in expected.bytes().zip(signature.bytes()) {
        result |= a ^ b;
    }
    result == 0
}

/// Verify webhook with signature and timestamp validation
pub fn verify(payload: &[u8], headers: &HashMap<String, String>, secret: &str, tolerance_seconds: Option<i64>) -> bool {
    let tolerance = tolerance_seconds.unwrap_or(DEFAULT_TOLERANCE_SECONDS);

    // Get signature from headers (case-insensitive)
    let signature = headers
        .get(SIGNATURE_HEADER)
        .or_else(|| headers.get(&SIGNATURE_HEADER.to_lowercase()))
        .map(String::as_str)
        .unwrap_or("");

    if signature.is_empty() {
        return false;
    }

    // Verify timestamp if present
    let timestamp_str = headers
        .get(TIMESTAMP_HEADER)
        .or_else(|| headers.get(&TIMESTAMP_HEADER.to_lowercase()));

    if let Some(ts_str) = timestamp_str {
        if let Ok(timestamp) = ts_str.parse::<i64>() {
            let now = Utc::now().timestamp();
            if (now - timestamp).abs() > tolerance {
                return false;
            }
        } else {
            return false;
        }
    }

    verify_signature(payload, signature, secret)
}

/// Construct and verify a webhook event
pub fn construct_event(
    payload: &[u8],
    headers: &HashMap<String, String>,
    secret: &str,
    tolerance_seconds: Option<i64>,
) -> Result<WebhookEvent> {
    let tolerance = if tolerance_seconds == Some(0) {
        DEFAULT_TOLERANCE_SECONDS
    } else {
        tolerance_seconds.unwrap_or(DEFAULT_TOLERANCE_SECONDS)
    };

    if !verify(payload, headers, secret, Some(tolerance)) {
        return Err(LinktorError::WebhookVerification {
            message: "Webhook signature verification failed".to_string(),
        });
    }

    let event: WebhookEvent = serde_json::from_slice(payload).map_err(|e| {
        LinktorError::WebhookVerification {
            message: format!("Failed to parse webhook event: {}", e),
        }
    })?;

    if event.id.is_empty() || event.event_type.is_empty() {
        return Err(LinktorError::WebhookVerification {
            message: "Invalid webhook event structure".to_string(),
        });
    }

    Ok(event)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_signature() {
        let payload = b"test payload";
        let secret = "test-secret";
        let signature = compute_signature(payload, secret);
        assert!(!signature.is_empty());
        assert_eq!(signature.len(), 64); // SHA256 produces 32 bytes = 64 hex chars
    }

    #[test]
    fn test_verify_signature() {
        let payload = b"test payload";
        let secret = "test-secret";
        let signature = compute_signature(payload, secret);
        assert!(verify_signature(payload, &signature, secret));
        assert!(!verify_signature(payload, "wrong-signature", secret));
    }
}
